### PR TITLE
Add IComponentActivator

### LINKED ISSUE #7962: Allow custom component instantiation
Change the access modifier of [ComponentFactory](https://github.com/aspnet/AspNetCore/blob/master/src/Components/Components/src/ComponentFactory.cs) to public to allow instatiation of components.  

An even better solution is to add an interface `IComponentFactory` that is implemented by `ComponentFactory` and that is registered in the `ServiceCollection`.  

[My current use case for this](https://github.com/AI4E/AI4E/blob/3fe96a1816146dbb8d9b12a54d6ae7883fdfce37/src/Blazor/AI4E.Blazor/Components/ViewExtension.cs#L162):
I implemented a system to split a blazor/razor-components project into multiple modules that are loaded lazily. The modules define view-extensions that can be rendered whereever needed. The view-extension rendering engines does not know about the concrete type of the component at compile time but needs to instantiate a component (of a type that is known at runtime), configure it with some values and finally render it. Currently I need to copy the source of `ComponentFactory` to my project to allow this. 

If you like, I can implement this and open a PR for it.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This is an updated version of PR #19642 by @stsrki. Please see the original PR for context and earlier discussion.

Fixes https://github.com/dotnet/aspnetcore/issues/7962

### CODE REVIEW

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/DefaultComponentActivator.cs:11]: CR note: this is `public` because people implementing a custom `IComponentActivator` may want to override the behavior in some cases, but fall back on the default behavior in other cases. We know the default behavior is just `Activator.CreateInstance`, but external developers don't necessarily know that, plus maybe we refine this behavior in some ways in the future.

**Comment by SteveSandersonMS** [comment]
Another CR note: this PR changes what was previously an invariant of the system. Until now, we always knew for sure that a rendertree frame of type "Component" with `ComponentType = T` for some type `T` would always correspond to a component instance of type `T`.

After this PR, that invariant no longer holds, since the activator could return a component instance of any `IComponent` type, not necessarily `T`.

Loosening this invariant is beneficial for cases like testing as it means you can "mock" a child component or do shallow rendering (replace all the child components with empty things). However it's something to bear in mind in the future and not get confused about!

File: src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs
Comment (javiercn) [inline src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs:110]: Does this need to be public? Channeling my inner rynowak, should we just have an Abstract Base Class here?

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (javiercn) [reply src/Components/Components/src/DefaultComponentActivator.cs:11]: If this is public we might as well get rid of the interface.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (javiercn) [inline src/Components/Components/src/RenderTree/Renderer.cs:77]: We can just register the default activator in DI.

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (pranavkm) [reply src/Components/Components/src/DefaultComponentActivator.cs:11]: Can we use an abstract base type + `Default` pattern:

```c#
public abstract class ComponentActivator
{
   public static ComponentActivator Default { get; } 
   public IComponent CreateInstance();
}
```

Making the implementation makes it harder to change it (for instance requires API reviews for an additional ctor etc)

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (pranavkm) [inline src/Components/Components/src/DefaultComponentActivator.cs:21]: Unrelated to your change, but it might be worth profiling this to see if caching the factories makes a difference.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:77]: I did that at first, but it’s more breaking. People who use Renderer outside our default DI config would then get errors. Example: bUnit. 

Do you have reasons for saying it’s important to require it to be in DI?

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs:11]: I’d rather not have any public static instance available, because then we’re locked into never having any state that you wouldn’t want to share across users. We can do static now, but we don’t know that will always be the case.

Dropping the interface and having a base class would be OK enough I think. So too  would just be keeping the interface and not exposing any concrete implementation publicly.

Do you have preferences among those options?

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs:21]: We only create one per Renderer so it seems very unlikely to make a difference to anything. Was there a particular reason you’re suspicious it might have an observable effect?

Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs:21]: Oh you mean some kind of cached activator? I’m not even sure that’s possible without Reflection.Emit (which works but is slow on WebAssembly). Do you know of a good way to implement it?

Comment (pranavkm) [reply src/Components/Components/src/DefaultComponentActivator.cs:21]: I was going to suggest ActivatorUtiltiies.CreateFactory but that uses Reflection.Emit. I guess once levi's ObjectFactory PR lands, we can revisit. Oh well 😞

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (pranavkm) [reply src/Components/Components/src/DefaultComponentActivator.cs:11]: We generally preference interfaces, so let's start with that. It's pretty trivial for users to replicate our code if they absolutely need to by copying it.

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs:21]: It's probably possible to do it like this:

```cs
internal interface ICachedActivator
{
    object CreateInstance();
}

internal class CachedActivator<T> : ICachedActivator where T: class, new()
{
    public object CreateInstance() => new T();
}
```

... and then separately create a `Dictionary<Type, ICachedActivator>` where for each `Type` we construct an instance one time using reflection.

However I have absolutely no reason to think this would be faster than `Activator.CreateInstance` when running on an interpreter. It's probably slower. With the interpreter, the only difference between `Activator.CreateInstance(typeof(T))` and `new T()` is having to look up the default constructor for `T`, which most likely takes less time than finding `typeof(T)` in a dictionary keyed on that. I don't know how `ObjectFactory` is implemented but it's hard to see how it would be faster than `Activator.CreateInstance` on an interpreter.

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/DefaultComponentActivator.cs]: OK, done

**Comment by SteveSandersonMS** [comment]
Pinging @dotnet/aspnet-blazor-eng for sign-off.

File: src/Components/Components/src/DefaultComponentActivator.cs
Comment (pranavkm) [reply src/Components/Components/src/DefaultComponentActivator.cs:15]: I think `new T()` uses Activator.CreateInstance under the covers: https://devblogs.microsoft.com/premier-developer/dissecting-the-new-constraint-in-c-a-perfect-example-of-a-leaky-abstraction/.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (pranavkm) [reply src/Components/Components/src/RenderTree/Renderer.cs:77]: Could we change this so that there's two ctors for Renderer, one that takes the default vs one without? It's convenient we have the serviceprovider, but having it be an explicit dependency makes it much more obvious and avoids yucky service location

**Review by pranavkm — APPROVED** [review]
Could you send an email over API-review? This API has precedence, since it's pretty close to how MVC's activator works. Outside of the few changes, this looks great.

File: src/Components/Components/src/RenderTree/Renderer.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:77]: Done

**Comment by stsrki** [comment]
@SteveSandersonMS With this PR merged I will finally be able to do what I wanted long time ago within my component library. I cannot thank you enough time for helping me. I really appreciate all your work. Thank you!

**Comment by egil** [comment]
@stsrki I am curious what you will use this feature for.

@SteveSandersonMS will this feature be added to the 3.x versions of Blazor as well, or will it only be available in the .net 5 version? Either way I guess I need to figure out how to do multi targeting correctly with bUnit.

**Comment by SteveSandersonMS** [comment]
It’s for 5.0 only. 

@stsrki Thanks for your help in making this happen too!

**Comment by stsrki** [comment]
@egil Basically in my components library [Blazorise](https://github.com/stsrki/Blazorise) every component in the root project can be overriden and customized by any the supported frameworks(Bootstrap, Material, AntDesign, etc.) But since untll now I didn't have freedom to create the components, I had to come up with some workarounds. Like, whenever any component is created and rendered for the first time I would check if there exists custom implementation. In case it exists it would render that one instead.

The problem was that I would basically create two components instead if just one. So with this PR I would be able to finally do that.

**Comment by egil** [comment]
@stsrki ahh I see. So for example, you have a public `<Button/>` component, which internally would have a `<BootstrapButton/>`, a `<MaterialButton/>`, etc. implementations, which you choose at runtime. I have done something similar before, just with differnet renderfragments depending on the context. I see how this will enable a more clean and simple approach.

I wonder if the public `<Button />` component in the scenario above could be an `interface` that inherits from `IComponent`. Would Blazor complain about this?

**Comment by stsrki** [comment]
@egil Yeah you got it. That's how it works. 

Well in theory it could be an interface but there is a lot more features and implementations behind the scenes. Just making it an interface would not suffice.

**Comment by egil** [comment]
@stsrki no I guess you have shared logic between them, so an interface might not be ideal. I was more wondering if it would be allowed by the Blazor compiler.

**Comment by springy76** [comment]
@egil using plain C# you can render both abstract and interface types (when handling those by implementing a capable `IComponentActivator`):

```c#
    private static RenderFragment RenderAbstractComponent(int arg1) => builder =>
    {
        builder.OpenComponent<IJustTestComponent>(1);
        builder.AddAttribute(2, nameof(IJustTestComponent.Arg1), arg1);
        builder.CloseComponent();
    };
```

It's only the razor interpreter which starts acting stupid as soon as the component class is abstract or an interface:

```razor
    <JustTestComponentBase Arg1="123" />
```

and lies on this with unnecessary false errors:
> error RZ10012: Found markup element with unexpected name 'JustTestComponentBase'.

This is frustating and imho if the referenced type finally implements `IComponent` the razor compiler should just leave it to the runtime as any other argument typo or argument-left-over just crashes at runtime when it doesn't match - or missing `ParameterAttribute` also just waits for the user at runtime to find the error.

**Pull request merged**
